### ISSUE TITLE
127.0.0.1 and localhost cannot be used interchangeably. 

### ISSUE THREAD

**Issue by SeanKilleen** [akkadotnet/akka.net#1186]
Per gitter conversation with @rogeralsing and @Aaronontheweb.

This gist has a problem in that messages will be dropped:

https://gist.github.com/SeanKilleen/9789908587de09cfe0f9

The reason for this is that I mix the concepts of `localhost` and `127.0.0.1` within the code. In the configuration, we define 127.0.0.1, but in the `RemoteScope` definition we use localhost.

Theoretically, these should be the same, but it results in messages not being delivered with no exception that makes this clear.

Not saying this is a huge issue, but it may bite first-time users who see those two addresses as conceptually equivalent.

**Comment by Aaronontheweb** [comment]
These values should not be interchangeable in Akka.NET - you bind an address to a hostname or IP, and that's how you address it. There are some reasons at the transport layer why this is a good idea, but the biggest reason is to enforce Akka.NET's concept of explicit choice.

However, from a DevOps perspective this is why the `public-hostname` config value exists in Akka.Remote:

```
akka
    { 
      log-config-on-start = on
      stdout-loglevel = INFO
      actor 
      {
        provider = "Akka.Remote.RemoteActorRefProvider, Akka.Remote"
      }
      remote {
        helios.tcp {
              transport-class =
           "Akka.Remote.Transport.Helios.HeliosTcpTransport, Akka.Remote"
              transport-protocol = tcp
              port = 8091
              hostname = "127.0.0.1"        
              public-hostname = "localhost"
              }
      }
    }
```

The reason why this is an issue is because **sockets bind to IP addresses and don't use DNS.** In JVM Akka, there's a 1:1 correlation between your socket address and the inbound endpoint of your actorsystem. If your transport can't resolve a hostname into an IP address that it has local access to on the calling machine, your transport can't bind.

So let's say for instance you bind a node to `0.0.0.0` (a special catch-all IP address) and it has a public hostname assigned externally ot that node through a cloud service of "foo.com." On that machine, Helios can bind to all network interfaces (i.e. if you have multiple network cards or virtual networks) and will receive input from all network devices.

If you send that system a message by addressing it by it's reachable hostname, "foo.com," that message*will* get delivered to the socket and Akka.Remote will receive it. However, Akka.Remote will see that this message was addressed to "foo.com" and NOT "0.0.0.0" and Akka.Remote will say "welp, not sure how this message got here - so I'm going to be safe by default and not pass it along."

You can resolve this issue by setting `public-hostname = "foo.com"` - this decouples your transports physical address binding from the logical address that external actor systems might use to send it messages.

TL;DR; because sockets :p

**Comment by SeanKilleen** [comment]
@Aaronontheweb fantastic explanation, thank you. Makes a lot of sense.

From the perspective of someone new to the framework, maybe we can include a note in the docs in the remoting section, a la "A quick note on hostnames vs IP Addresses". I think we could sum this up as "you may think that 127.0.0.1 and localhost are the same thing, but this is not a guarantee. Make sure these agree in your config if you're doing remoting." 

If we think that this is worthwhile for others (vs. Sean-dunces only :wink: ) I can add a section to the docs.

**Comment by Aaronontheweb** [comment]
> If we think that this is worthwhile for others (vs. Sean-dunces only :wink: ) I can add a section to the docs.

Yeah, this is an issue of understanding how low-level socket programming stuff impacts the way you use Akka.NET over TCP / UDP. I'd be **great** if you could add something to the docs about that. We have virtually nothing on Akka.Remote today. Would you mind adding a new page / folder for Akka.Remote configuration on http://getakka.net/ ?

**Comment by SeanKilleen** [comment]
Sure thing. I'll leave this here until I get an issue created in the appropriate repo that can refer to it.

**Comment by skotzko** [comment]
@Seankilleen great one to bring up. You are *definitely* not the only person who has thought this (*raises hand*)...

Also, @aaronontheweb and I are going to be writing a bunch of docs on remote/cluster over the next few days so let's sync up about that and make sure we aren't duplicating work.
